aten/src/ATen/templates/Functions.h

```diff
@@ -28,6 +28,7 @@ inline Tensor from_blob(
     IntArrayRef strides,
     const std::function<void(void*)>& deleter,
     const TensorOptions& options = {}) {
+  AutoNonVariableTypeMode guard;
```

Comment:
this seems "wrong" in the sense that I would expect AutoNonVariableTypeMode only at the torch level.  I.e. I think this is actually correct, but not maintainable, this should be going through variable_factories.

Reply:
`from_blob` is special because there's no torch level binding code for it, and it's manually bound in TensorFunctions. So there isn't anywhere else to put it: user level C++ calls will directly drop here. So yes, it's "unmaintainable", but only in the same sense all of the manually bound functions in PyTorch are unmaintainable. It's something you'll have to remember to do, along with setting device guards, etc. Probably something we should talk about in the docs.

One question is if plans to separate tracing into its own mode help make this more maintainable. I don't think it does though, for precisely the reason that it is manually bound.

Reply:
But why is it manually bound?  We have a bunch of related support at the torch level for things like this (e.g. torch::tensor) -- what's special about this one?

Reply:
It looks like torch::tensor is manually bound too. Actually, the situation in torch::tensor is quite similar, except that the call to AutoNonVariableTypeMode happens inside torch namespace rather than at namespace.

`torch::tensor` is defined in tools/autograd/templates/variable_factories.h and the action happens in `torch/csrc/api/include/torch/detail/TensorDataContainer.h`. There isn't really any reason this class has to live in torch versus at; it is operating at basically the same level of abstraction as `from_blob`